pennylane/gradients/parameter_shift.py

```diff
@@ -671,6 +671,16 @@ def var_param_shift(tape, argnum, shifts=None, gradient_recipes=None, f0=None, b
     for i in var_indices:
         obs = expval_tape.measurements[i].obs
         expval_tape._measurements[i] = qml.expval(op=obs)
+        if obs.name == "Hamiltonian":
+            first_obs_idx = len(expval_tape.operations)
+            for t_idx in reversed(range(len(expval_tape.trainable_params))):
+                op, op_idx, _ = expval_tape.get_operation(t_idx)
+                if op_idx < first_obs_idx:
+                    break  # already seen all observables
+                if op is obs:
+                    raise ValueError(
+                        "Can only differentiate Hamiltonian coefficients for expectations, not variances"
+                    )

```

Comment:
Is this a new constraint? Or an already existing one that we didn't catch properly?

Reply:
previously existing. see the point on `var_param_shift` in the PR description for why we didn't need to check it here before.